fasjson_client/cli/get_cert.py

```diff
@@ -0,0 +1,248 @@
+
+    Returns:
+        x509.Certificate: the existing certificate
+    """
+    try:
+        client = Client(url)
+    except ClientError as e:
+        raise click.ClickException("could not get existing certificate ({}).".format(e))
+
+    try:
+        response = client.get_user(username=username)
+    except APIError as e:
+        if e.code == 404:
+            raise click.ClickException("user {} not found.".format(username))
+        else:
+            raise click.ClickException(str(e))
+
+    certificates = response.result["certificates"]
+    if not certificates:
+        return None
+    certificates = [_load_cert_from_base64(cert) for cert in certificates]
+    # Get the latest certificate. Don't sort by serial_number, IPA may switch to randomized serial
+    # numbers in the future to avoid the MD5 collision attack.
+    certificates.sort(key=lambda c: c.not_valid_after)
+    return certificates[-1]
+
+
+def _write_certificate(cert, path):
+    """Write the certificate to the filesystem in PEM format
+
+    Args:
+        cert (x509.Certificate): the certificate to write
+        path (str): the filesystem path to write to
+    """
+    cert = cert.public_bytes(encoding=serialization.Encoding.PEM)
+    with open(path, "wb") as f:
+        f.write(cert)
+
+
+@click.option(
+    "-u", "--username", default=None, show_default=True, help="Your FAS username.",
+)
+@click.option(
+    "-p",
+    "--private-key",
+    type=click.Path(),
+    help="The path to the private key. If it does not exist, it will be generated.",
+)
+@click.option(
+    "-s",
+    "--save-to",
+    required=True,
+    type=click.Path(),
+    help="The path to save your certificate to.",
+)
+@click.option(
+    "--existing",
+    is_flag=True,
+    default=False,
+    help="Retrieve an existing certificate instead of generating a new one.",
+)
+@click.pass_context
+def get_cert(ctx, username, private_key, save_to, existing):
+    """Get a certificate for the provided username.
+
+    The certificate can be an existing one or a new one. In this case a private key is either
+    required or generated.
+    """
+    if os.path.exists(save_to):
+        raise click.BadParameter(
+            "the destination file {} already exists. Aborting.".format(save_to),
+            param_hint="save_to",
+        )
+    if username is None:
+        try:
+            username = getuser()
+        except Exception:
+            raise click.BadParameter(
+                "you must provide a username.", param_hint="username"
+            )
+
+    url = ctx.obj["url"]
+
+    if existing:
+        log.debug("Looking for an existing certificate...")
+        cert = _get_existing(username, url)
+        if cert is None:
+            raise click.ClickException(
+                "No existing certificate, you need to request one."
+            )
+        _write_certificate(cert, save_to)
+        log.info("Certificate written to {}".format(save_to))
+        return
+
+    if not private_key:
```

Comment:
Shouldn't this also depend on `save_to` and/or `existing`? As it is if you don't provide the private key then it will never get passed this point, correct?

Reply:
If `existing` is true, then you are in the case handled just above and you'll never reach this test.
The `save_to` parameter is marked as required so it'll have to be set.
I'm not sure I understand what case you're pointing out.

Reply:
Yes sorry, I misunderstood the params.